src/api/cubrid_log.c

```diff
@@ -0,0 +1,1533 @@
+
+  for (i = 0; i < num_infos; i++)
+    {
+      if ((rc = cubrid_log_make_log_item (&ptr, &g_log_items[i]) != CUBRID_LOG_SUCCESS))
+	{
+	  CUBRID_LOG_ERROR_HANDLING (rc, trace_errbuf);
+	}
+
+      g_log_items[i].next = &g_log_items[i + 1];
+
+      ptr = PTR_ALIGN (ptr, MAX_ALIGNMENT);
+    }
+
+  g_log_items[num_infos - 1].next = NULL;
+
+  *log_item_list = g_log_items;
+  *list_size = num_infos;
+
+  return CUBRID_LOG_SUCCESS;
+
+cubrid_log_error:
+
+  return err_code;
+}
+
+/*
+ * cubrid_log_extract () -
+ *   return:
+ *   lsa(in/out):
+ *   log_item_list(out):
+ *   list_size(out):
+ */
+int
+cubrid_log_extract (uint64_t * lsa, CUBRID_LOG_ITEM ** log_item_list, int *list_size)
+{
+  int num_infos, total_length;
+  int err_code;
+  int rc;
+
+  char trace_errbuf[1024];
+
+  if (g_stage != CUBRID_LOG_STAGE_PREPARATION && g_stage != CUBRID_LOG_STAGE_EXTRACTION)
+    {
+      CUBRID_LOG_ERROR_HANDLING (CUBRID_LOG_INVALID_FUNC_CALL_STAGE, trace_errbuf);
+    }
+
+  if (lsa == NULL || log_item_list == NULL || list_size == NULL)
+    {
+      CUBRID_LOG_ERROR_HANDLING (CUBRID_LOG_INVALID_OUT_PARAM, trace_errbuf);
+    }
+
+  memcpy (&g_next_lsa, lsa, sizeof (LOG_LSA));
+
+  rc = cubrid_log_extract_internal (&g_next_lsa, &num_infos, &total_length);
+
+  if (rc != CUBRID_LOG_SUCCESS)
+    {
+      CUBRID_LOG_ERROR_HANDLING (rc, trace_errbuf);
+    }
+
+  if ((rc = cubrid_log_make_log_item_list (num_infos, total_length, log_item_list, list_size)) != CUBRID_LOG_SUCCESS)
+    {
+      CUBRID_LOG_ERROR_HANDLING (rc, trace_errbuf);
+    }
+
+  memcpy (lsa, &g_next_lsa, sizeof (uint64_t));
+
+  g_stage = CUBRID_LOG_STAGE_EXTRACTION;
+
+  return CUBRID_LOG_SUCCESS;
+
+cubrid_log_error:
+
+  return err_code;
+}
+
+static int
+cubrid_log_clear_data_item (DATA_ITEM_TYPE data_item_type, CUBRID_DATA_ITEM * data_item)
+{
+  int err_code;
+
+  switch (data_item_type)
+    {
+    case DATA_ITEM_TYPE_DDL:
+      /* nothing to do */
+      break;
+
+    case DATA_ITEM_TYPE_DML:
+      free (data_item->dml.changed_column_index);
+      free (data_item->dml.changed_column_data);
+      free (data_item->dml.changed_column_data_len);
+
+      free (data_item->dml.cond_column_index);
+      free (data_item->dml.cond_column_data);
+      free (data_item->dml.cond_column_data_len);
+
+      break;
+
+    case DATA_ITEM_TYPE_DCL:
+      /* nothing to do */
+      break;
+
+    case DATA_ITEM_TYPE_TIMER:
+      /* nothing to do */
+      break;
+
+    default:
+      assert (0);
+    }
+
+  return CUBRID_LOG_SUCCESS;
+
+cubrid_log_error:
+
+  return err_code;
+}
+
+/*
+ * cubrid_log_clear_log_item () -
+ *   return:
+ *   log_item_list(in):
+ */
+int
+cubrid_log_clear_log_item (CUBRID_LOG_ITEM * log_item_list)
+{
+  int i;
+  int err_code;
+  int rc;
+
+  char trace_errbuf[1024];
+
+  if (g_stage != CUBRID_LOG_STAGE_EXTRACTION)
+    {
+      CUBRID_LOG_ERROR_HANDLING (CUBRID_LOG_INVALID_FUNC_CALL_STAGE, trace_errbuf);
+    }
+
+  if (log_item_list == NULL)
```

Comment:
The address of the log_item_list is equaled to g_log_items. So, It needs to check.